### ISSUE TITLE
any plan to add openssl mem bio?

### ISSUE THREAD

**Issue by rainbowvalue** [cl-plus-ssl/cl-plus-ssl#184]
hello, i want to use mem_bio, i do not need to use with socket fd or io stream. i just need mem_buf. 
;)
please help, thank you.

https://www.openssl.org/docs/man3.0/man3/BIO_s_mem.html

**Comment by avodonosov** [comment]
No such plans.

What help do you need? Look how make-ssl-client-stream is implemented, and do something similar. Define an ssl wrapper for BIO_s_mem / BIO_s_secmem.

If you'll have something others might benefit from, pull requests are welcome.

**Comment by rainbowvalue** [comment]
>  an ssl wrapper for BIO_s_mem

yes, that is what i need.

**Comment by avodonosov** [comment]
@rainbowvalue, have you succeeded? What is your use case?

**Comment by rainbowvalue** [comment]
There is no success for the time being, I am looking at the documentation recently.

I plan to create two mem bufs (buf1,buf2) for each non-blocking socket.

write data1 of recv socket to buf1, read data2 from buf2 and send socket.

---
My vision of non blocking socket is：
|function|len=-1|len=0|len>0 (n)|
|:--|:--|:--|:--|
|recv|socket has closed (error)|no data (recv later)|had recv n length octets (finish)|
|send|socket has closed (error)|full buffer (send later)|had send n length octets (finish)|

**Comment by avodonosov** [comment]
@rainbowvalue, mem bio is not needed for nonblocking IO.

See the docs for SSL_write and SSL_read. Usual socket BIO can be used, if the socket is in non-blocking mode. 

SSL_write and SSL_read, in case the underlying BIO is nonblocking, as happens with socket BIO with nonblocking sockets, the functions can finish with SSL_get_error() returning SSL_ERROR_WANT_READ / SSL_ERROR_WANT_WRITE, which means the application should wait for socket to be ready for read or write (using select, poll, whatever), and retry the call.

That way one can implement non-blocking IO.

cl+ssl even uses this possibility, but only to support timeouts on SBCL and CCL.

See function [ensure-ssl-funcall](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/17d5cdd65405f1d26e26f3e875e70027d0c8eedb/src/ssl-funcall.lisp#L140C8-L140C26).
This function wraps calls to OpenSSL IO functions, like SSL_read and SSL_write (passed through the parameter func) to perform some error handling. As you see, in case of SSL_ERROR_WANT_READ or SSL_ERROR_WANT_WRITE it waits for socket to become available and repeats the call. In this case the waiting is synchronous, and wait time is limited by possible deadline / timeout. 

So in order to implement really non-blocking IO (single thread handling multiple sockets), one can do a similar thing, only instead of synchronous waiting on the current socket in the same thread, add the socket to a select list for the main IO thread.

cl+ssl even has a function https://github.com/cl-plus-ssl/cl-plus-ssl/blob/17d5cdd65405f1d26e26f3e875e70027d0c8eedb/src/ssl-funcall.lisp#L161 which simply returns instead of waiting and retry. I have never used it, and maybe it's not super convenient, because it does not return result of ssl-get-error to the caller. See yourself if this function is useful for you or you need to implement something similar yourself.

**Comment by rainbowvalue** [comment]
> cl+ssl even has a function [nonblocking-ssl-funcall](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/17d5cdd65405f1d26e26f3e875e70027d0c8eedb/src/ssl-funcall.lisp#L161) which simply returns instead of waiting and retry. 

> See also [install-nonblock-flag](https://github.com/cl-plus-ssl/cl-plus-ssl/blob/17d5cdd65405f1d26e26f3e875e70027d0c8eedb/src/streams.lisp#L190)

Thanks, these helped me a lot, I will try them.
